src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParsing.scala

```diff
@@ -0,0 +1,302 @@
+    parse(expression, attributeExpression)
+  }
+
+  def parseOutputExpr(expression: String) = {
+    parse(expression, outputExpression)
+  }
+
+  def parseEntityExpr(expression: String) = {
+    parse(expression, entityExpression)
+  }
+
+  private def parse(expression: String, parser: Parser[PipelineQuery] ) = {
+    //Attempt to parse the expression into a pipeline query to hand off to Gremlin
+    parseAll(parser, expression) match {
+      case Success(result, _) => scala.util.Success(result)
+      case NoSuccess(msg, next) => scala.util.Failure(new RuntimeException("Failed at line %s, column %s: %s".format(next.pos.line, next.pos.column, msg)))
+    }
+  }
+
+  /** functions against pipes **/
+
+  // the root function starts the pipeline at some root entity type in the workspace
+  private def entityRootFunc(context: SlickExpressionContext): PipeType = {
+    for {
+      entity <- entityQuery.findEntityByName(context.workspaceContext.workspaceId, context.rootType, context.rootName)
+    } yield (entity, 0)
+  }
+
+  // final func that gets an attribute off a workspace
+  private def workspaceAttributeRootFunc(attributeName: String)(context: SlickExpressionContext, graphPipeline: Option[PipeType]): ReadAction[Iterable[Attribute]] = {
+    (for {
+      workspace <- workspaceQuery.findByIdQuery(context.workspaceContext.workspaceId)
+      workspaceAttribute <- workspaceAttributeQuery if workspaceAttribute.workspaceId === workspace.id
+      attribute <- attributeQuery if workspaceAttribute.attributeId === attribute.id && attribute.name === attributeName
+    } yield attribute).result.map(x => Seq(attributeQuery.unmarshalAttributes(x.map((_, None))).getOrElse(attributeName, AttributeNull)))
+  }
+
+  // root func that gets an entity reference off a workspace
+  private def workspaceEntityRefRootFunc(entityRefName: String)(context: SlickExpressionContext): PipeType = {
+    for {
+      workspace <- workspaceQuery.findByIdQuery(context.workspaceContext.workspaceId)
+      workspaceAttribute <- workspaceAttributeQuery if workspaceAttribute.workspaceId === workspace.id
+      attribute <- attributeQuery if workspaceAttribute.attributeId === attribute.id && attribute.name === entityRefName
+      nextEntity <- entityQuery if attribute.valueEntityRef === nextEntity.id
+    } yield (nextEntity, attribute.listIndex.getOrElse(0))
+  }
+
+  // add pipe to an entity referenced by the current entity
+  private def entityNameAttributePipeFunc(entityRefName: String)(context: SlickExpressionContext, queryPipeline: PipeType): PipeType = {
+      for {
+      (entity, listIndex) <- queryPipeline
+      entityAttribute <- entityAttributeQuery if entity.id === entityAttribute.entityId
+      attribute <- attributeQuery if entityAttribute.attributeId === attribute.id && attribute.name === entityRefName
+      nextEntity <- entityQuery if attribute.valueEntityRef === nextEntity.id
+    } yield (nextEntity, listIndex * 10 + attribute.listIndex.getOrElse(0))
```

Comment:
that seems like the logical choice. i don't think there's any need to preserve ordering across tests as long as it's internally consistent